File: src/Scada.Engineering/VisualScripting/LegacyVisualEngineeringPropertyCodec.cs
using System.Collections.ObjectModel;
using System.Globalization;
using System.Text.RegularExpressions;

namespace Scada.Engineering.VisualScripting;

/// <summary>
/// Transitional codec between the schema-v10/v11 string property bag and the
/// typed visual property model. Conversion is always driven by the declared
/// VisualObjectPropertySchema; values are never type-guessed from text.
///
/// This exists to keep the graphical editor and Runtime on one conversion path
/// until a later schema revision can store typed visual values directly.
/// </summary>
public static class LegacyVisualEngineeringPropertyCodec
{
    private static readonly Regex CanonicalInteger = new(
        @"^-?(?:0|[1-9][0-9]*)$",
        RegexOptions.CultureInvariant);

    private static readonly Regex CanonicalNumber = new(
        @"^-?(?:0|[1-9][0-9]*)(?:\.[0-9]+)?(?:[eE][+-]?[0-9]+)?$",
        RegexOptions.CultureInvariant);

    public static IReadOnlyDictionary<string, VisualPropertyValue> Decode(
        VisualObjectPropertySchema schema,
        IReadOnlyDictionary<string, string>? serializedValues)
    {
        ArgumentNullException.ThrowIfNull(schema);

        var decoded = new Dictionary<string, VisualPropertyValue>(StringComparer.Ordinal);
        foreach (var pair in serializedValues ?? new Dictionary<string, string>())
        {
            var definition = schema.GetRequired(pair.Key);
            var value = DecodeValue(definition, pair.Value);
            definition.ValidateValue(value);
            decoded.Add(pair.Key, value);
        }

        return new ReadOnlyDictionary<string, VisualPropertyValue>(decoded);
    }

    public static IReadOnlyDictionary<string, string> Encode(
        VisualObjectPropertySchema schema,
        IReadOnlyDictionary<string, VisualPropertyValue>? typedValues)
    {
        ArgumentNullException.ThrowIfNull(schema);

        var encoded = new Dictionary<string, string>(StringComparer.Ordinal);
        foreach (var pair in typedValues ?? new Dictionary<string, VisualPropertyValue>())
        {
            var definition = schema.GetRequired(pair.Key);
            definition.ValidateValue(pair.Value);

            var serialized = EncodeValue(pair.Value);
            if (serialized is not null)
                encoded.Add(pair.Key, serialized);
        }

        return new ReadOnlyDictionary<string, string>(encoded);
    }

    private static VisualPropertyValue DecodeValue(VisualPropertyDefinition definition, string serialized)
    {
        ArgumentNullException.ThrowIfNull(serialized);

        return definition.ValueKind switch
        {
            VisualPropertyValueKind.Boolean => DecodeBoolean(definition.Key, serialized),
            VisualPropertyValueKind.Number => DecodeNumber(definition.Key, serialized),
            VisualPropertyValueKind.Integer => DecodeInteger(definition.Key, serialized),
            VisualPropertyValueKind.String => new VisualStringValue(serialized),
            VisualPropertyValueKind.Color => new VisualColorValue(serialized),
            VisualPropertyValueKind.ResourceReference => new VisualResourceReferenceValue(serialized),
            VisualPropertyValueKind.AssetReference => new VisualAssetReferenceValue(serialized),
            _ => throw new InvalidDataException(
                $"Unsupported visual property value kind '{definition.ValueKind}' for '{definition.Key}'.")
        };
    }

    private static VisualPropertyValue DecodeBoolean(string propertyKey, string serialized)
    {
        if (serialized == "true") return new VisualBooleanValue(true);
        if (serialized == "false") return new VisualBooleanValue(false);

        throw new InvalidDataException(
            $"Visual property '{propertyKey}' requires canonical boolean text 'true' or 'false'.");
    }

    private static VisualPropertyValue DecodeNumber(string propertyKey, string serialized)
    {
        if (!CanonicalNumber.IsMatch(serialized) ||
            !double.TryParse(
                serialized,
                NumberStyles.Float,
                CultureInfo.InvariantCulture,
                out var value) ||
            double.IsNaN(value) ||
            double.IsInfinity(value))
        {
            throw new InvalidDataException(
                $"Visual property '{propertyKey}' requires a canonical finite invariant-culture number.");
        }

        return new VisualNumberValue(value);
    }

    private static VisualPropertyValue DecodeInteger(string propertyKey, string serialized)
    {
        if (!CanonicalInteger.IsMatch(serialized) ||
            !int.TryParse(
                serialized,
                NumberStyles.AllowLeadingSign,
                CultureInfo.InvariantCulture,
                out var value))
        {
            throw new InvalidDataException(
                $"Visual property '{propertyKey}' requires a canonical 32-bit invariant-culture integer.");
        }

        return new VisualIntegerValue(value);
    }

    private static string? EncodeValue(VisualPropertyValue value) => value switch
    {
        VisualBooleanValue boolean => boolean.Value ? "true" : "false",
        VisualNumberValue number => number.Value.ToString("R", CultureInfo.InvariantCulture),
        VisualIntegerValue integer => integer.Value.ToString(CultureInfo.InvariantCulture),
        VisualStringValue text => text.Value,
        VisualColorValue color => color.Value,
        VisualResourceReferenceValue resource => resource.ResourceId,
        VisualAssetReferenceValue { AssetId: not null } asset => asset.AssetId,
        VisualAssetReferenceValue => null,
        _ => throw new InvalidDataException(
            $"Unsupported visual property value kind '{value.Kind}'.")
    };
}

File: web/scada-web/src/visual-runtime/legacyVisualEngineeringPropertyCodec.ts
import { VisualObjectPropertySchema } from './visualPropertyRegistry';
import {
  cloneVisualPropertyValue,
  VisualPropertyContractError,
  type AssetReference,
  type VisualPropertyDefinition,
  type VisualPropertyValue
} from './visualPropertyTypes';

const CANONICAL_INTEGER = /^-?(?:0|[1-9][0-9]*)$/;
const CANONICAL_NUMBER = /^-?(?:0|[1-9][0-9]*)(?:\.[0-9]+)?(?:[eE][+-]?[0-9]+)?$/;
const INT32_MIN = -2147483648;
const INT32_MAX = 2147483647;

/**
 * Transitional adapter for schema-v10/v11 string-valued visual Engineering.
 * Conversion is directed exclusively by the declared property schema. No value
 * is type-guessed from its textual contents.
 */
export function decodeLegacyVisualEngineeringProperties(
  serializedValues: Readonly<Record<string, string>> | null | undefined,
  schema: VisualObjectPropertySchema
): Readonly<Record<string, VisualPropertyValue>> {
  const decoded: Record<string, VisualPropertyValue> = Object.create(null) as Record<string, VisualPropertyValue>;

  for (const [propertyKey, serialized] of Object.entries(serializedValues ?? {})) {
    const definition = schema.getRequired(propertyKey);
    const candidate = decodeValue(definition, serialized);
    const validation = schema.validate(propertyKey, candidate);
    if (!validation.ok) {
      throw new VisualPropertyContractError(
        `legacyProperty.${validation.code}`,
        `Legacy Engineering value for '${propertyKey}' is invalid: ${validation.code}.`,
        propertyKey
      );
    }
    decoded[propertyKey] = cloneVisualPropertyValue(validation.value);
  }

  return Object.freeze(decoded);
}

/**
 * Encodes explicitly engineered typed values back into the legacy string bag.
 * A null assetRef is represented by absence of the property, not by a sentinel.
 */
export function encodeLegacyVisualEngineeringProperties(
  typedValues: Readonly<Record<string, VisualPropertyValue>> | null | undefined,
  schema: VisualObjectPropertySchema
): Readonly<Record<string, string>> {
  const encoded: Record<string, string> = Object.create(null) as Record<string, string>;

  for (const [propertyKey, candidate] of Object.entries(typedValues ?? {})) {
    const definition = schema.getRequired(propertyKey);
    const validation = schema.validate(propertyKey, candidate);
    if (!validation.ok) {
      throw new VisualPropertyContractError(
        `legacyProperty.${validation.code}`,
        `Typed Engineering value for '${propertyKey}' is invalid: ${validation.code}.`,
        propertyKey
      );
    }

    const serialized = encodeValue(definition, validation.value);
    if (serialized !== null) encoded[propertyKey] = serialized;
  }

  return Object.freeze(encoded);
}

function decodeValue(definition: VisualPropertyDefinition, serialized: string): VisualPropertyValue {
  if (typeof serialized !== 'string') {
    throw invalidSerialized(definition.key, 'Legacy visual property values must be strings.');
  }

  switch (definition.type) {
    case 'boolean':
      if (serialized === 'true') return true;
      if (serialized === 'false') return false;
      throw invalidSerialized(definition.key, "Boolean text must be exactly 'true' or 'false'.");

    case 'number': {
      if (definition.integer === true) {
        if (!CANONICAL_INTEGER.test(serialized)) {
          throw invalidSerialized(definition.key, 'Integer text is not canonical.');
        }
        const value = Number(serialized);
        if (!Number.isSafeInteger(value) || value < INT32_MIN || value > INT32_MAX) {
          throw invalidSerialized(definition.key, 'Integer is outside the supported 32-bit range.');
        }
        return value;
      }

      if (!CANONICAL_NUMBER.test(serialized)) {
        throw invalidSerialized(definition.key, 'Number text is not canonical.');
      }
      const value = Number(serialized);
      if (!Number.isFinite(value)) {
        throw invalidSerialized(definition.key, 'Number must be finite.');
      }
      return value;
    }

    case 'string':
    case 'color':
    case 'enum':
      return serialized;

    case 'assetRef':
      return Object.freeze({ assetId: serialized } satisfies AssetReference);
  }
}

function encodeValue(
  definition: VisualPropertyDefinition,
  value: VisualPropertyValue
): string | null {
  switch (definition.type) {
    case 'boolean':
      return value === true ? 'true' : 'false';

    case 'number':
      return String(value as number);

    case 'string':
    case 'color':
    case 'enum':
      return value as string;

    case 'assetRef':
      return value === null ? null : (value as AssetReference).assetId;
  }
}

function invalidSerialized(propertyKey: string, detail: string): VisualPropertyContractError {
  return new VisualPropertyContractError(
    'legacyProperty.invalidSerializedValue',
    `Legacy Engineering property '${propertyKey}' is invalid. ${detail}`,
    propertyKey
  );
}
